File: src/main/java/com/rust/util/Camera.java
package com.rust.util;

import glm.Glm;
import glm.mat._4.Mat4;
import glm.vec._3.Vec3;
import glm.vec._4.Vec4;

import static java.lang.Math.cos;
import static java.lang.Math.sin;

public class Camera {
    public static float YAW = -90.0f;
    public static float PITCH = 0.0f;
    public static float SPEED = 100.0f;
    public static float SENSITIVITY = 0.8f;
    public static float ZOOM = 45.0f;
    // camera Attributes
    public Vec3 Position;
    // object center
    public Vec3 Center;
    public Vec3 Front;
    public Vec3 Up;
    public Vec3 Right;
    public Vec3 WorldUp;
    public Vec3 WorldDown;
    // euler Angles
    float Yaw;
    float Pitch;
    // camera options
    float MovementSpeed;
    float MouseSensitivity;
    public float Zoom;
    public Camera(){
        Front = new Vec3(0,0,-1);
        MovementSpeed = SPEED;
        MouseSensitivity = SENSITIVITY;
        Zoom = ZOOM;
        Position = new Vec3(0,0,0);
        WorldUp = new Vec3(0,1,0);
        WorldDown = new Vec3(0,-1,0);
        Center = new Vec3(0,0,0);
        Yaw = YAW;
        Pitch = PITCH;
        updateCameraVectors();
    }
    // constructor with vectors
    public Camera(Vec3 position, Vec3 up, float yaw, float pitch)
    {
        Front = new Vec3(0,0,-1);
        MovementSpeed = SPEED;
        MouseSensitivity = SENSITIVITY;
        Zoom = ZOOM;
        Position = position;
        WorldUp = up;
        Yaw = yaw;
        Pitch = pitch;
        updateCameraVectors();
    }
    // constructor with scalar values
    public Camera(float posX, float posY, float posZ, float upX, float upY, float upZ, float yaw, float pitch)
    {
        Front = new Vec3(0,0,-1);
        MovementSpeed = SPEED;
        MouseSensitivity = SENSITIVITY;
        Zoom = ZOOM;
        Position = new Vec3(posX, posY, posZ);
        WorldUp = new Vec3(upX, upY, upZ);
        Yaw = yaw;
        Pitch = pitch;
        updateCameraVectors();
    }

    public Mat4 GetViewMatrix()
    {
    // returns the view matrix calculated using Euler Angles and the LookAt Matrix
        Vec3 center = GlmExtension.add(Position,Front);
        return GlmExtension.lookAtRH(Position,center,Up);
    }
    public void FrameFun(){
        clocker += TimeManager.shared.deltaTime;
    }
    private float clocker = 0f;
    // processes input received from any keyboard-like input system. Accepts input parameter in the form of camera defined ENUM (to abstract it from windowing systems)
    public void ProcessKeyboard(Camera_Movement direction, float deltaTime)
    {
        float velocity = MovementSpeed * deltaTime;
        if (direction == Camera_Movement.FORWARD)
            camera_rotate(true,false,deltaTime);
        if (direction == Camera_Movement.BACKWARD)
            camera_rotate(false,false,deltaTime);
        if (direction == Camera_Movement.LEFT)
            camera_rotate(true,true,deltaTime);
        if (direction == Camera_Movement.RIGHT)
            camera_rotate(false,true,deltaTime);/*
        float velocity = MovementSpeed * deltaTime / 10;
        if (direction == Camera_Movement.FORWARD)
            Position = GlmExtension.sub(Position,new Vec3(0,0,velocity));
        if (direction == Camera_Movement.BACKWARD)
            Position = GlmExtension.add(Position,new Vec3(0,0,velocity));
        if (direction == Camera_Movement.LEFT)
            Position = GlmExtension.sub(Position,new Vec3(velocity,0,0));
        if (direction == Camera_Movement.RIGHT)
            Position = GlmExtension.add(Position,new Vec3(velocity,0,0));*/
    }
    private void camera_rotate(boolean isClockWise, boolean isHorizontal , float deltaTime){
        float delta = MovementSpeed * deltaTime;
        Vec3 t = GlmExtension.sub(Position,Center);
        Vec4 currentDelta = new Vec4(t.x,t.y,t.z,1);
        Mat4 rotate = new Mat4(1);
        if(isClockWise){
            if (isHorizontal){
                rotate.rotate(GlmExtension.radians(delta),0,1,0);
            }else{
                rotate.rotate(GlmExtension.radians(delta),Right.x,Right.y,Right.z);
            }
            currentDelta = rotate.mul(currentDelta);
            Position = GlmExtension.add(Center , new Vec3(currentDelta.x,currentDelta.y,currentDelta.z));
            Vec4 t2 = new Vec4(Front.x,Front.y,Front.z,0);
            t2 = rotate.mul(t2);
            Front = new Vec3(t2.x,t2.y,t2.z);
            if (GlmExtension.sub(Front,WorldUp).length()>0.1f){
                Right = Glm.cross_(Front, WorldUp).normalize();
            }else{
                Right = Glm.cross_(Front, WorldDown).normalize();
            }
        }else {
            if(isHorizontal){
                rotate.rotate(GlmExtension.radians(-delta),0,1,0);
            }else{
                rotate.rotate(GlmExtension.radians(-delta),Right.x,Right.y,Right.z);
            }
            currentDelta = rotate.mul(currentDelta);
            Position = GlmExtension.add(Center , new Vec3(currentDelta.x,currentDelta.y,currentDelta.z));
            Vec4 t2 = new Vec4(Front.x,Front.y,Front.z,0);
            t2 = rotate.mul(t2);
            Front = new Vec3(t2.x,t2.y,t2.z);
            if (GlmExtension.sub(Front,WorldUp).length()>0.1f){
                Right = Glm.cross_(Front, WorldUp).normalize();
            }else{
                Right = Glm.cross_(Front, WorldDown).normalize();
            }
        }
    }
    // processes input received from a mouse input system. Expects the offset value in both the x and y direction.
    public void ProcessMouseMovement(double xoffset, double yoffset, boolean constrainPitch)
    {
        xoffset *= MouseSensitivity;
        yoffset *= MouseSensitivity;
        Yaw += xoffset;
        Pitch += yoffset;
        // make sure that when pitch is out of bounds, screen doesn't get flipped
        if (constrainPitch)
        {
            if (Pitch > 89.0f)
                Pitch = 89.0f;
            if (Pitch < -89.0f)
                Pitch = -89.0f;
        }
        // update Front, Right and Up Vectors using the updated Euler angles
        updateCameraVectors();
    }

    // processes input received from a mouse scroll-wheel event. Only requires input on the vertical wheel-axis
    public void ProcessMouseScroll(float yoffset)
    {
        Zoom -= (float)yoffset;
        if (Zoom < 1.0f)
            Zoom = 1.0f;
        if (Zoom > 45.0f)
            Zoom = 45.0f;
    }
    // calculates the front vector from the Camera's (updated) Euler Angles
    public void updateCameraVectors()
    {
        // calculate the new Front vector
        Vec3 front = new Vec3();
        front.x = (float) (cos(radians(Yaw)) * cos(radians(Pitch)));
        front.y = (float) sin(radians(Pitch));
        front.z = (float) (sin(radians(Yaw)) * cos(radians(Pitch)));
        Front = front.normalize_();
        // also re-calculate the Right and Up vector
        Right = Glm.cross_(Front, WorldUp).normalize();  // normalize the vectors, because their length gets closer to 0 the more you look up or down which results in slower movement.
        Up = Glm.cross_(Right, Front).normalize();
    }
    public static double radians(double degrees){
        return degrees * 0.01745329251994329576923690768489;
    }
}

File: Camera.cpp
#include "Camera.h"
float Camera::YAW = -95;
float Camera::PITCH = 0;
float Camera::SPEED = 1;
float Camera::SENSITIVITY = 0.8f;
float Camera::ZOOM = 10;
Camera::Camera()
{
    Front = vec3(0, 0, -1);
    MovementSpeed = SPEED;
    MouseSensitivity = SENSITIVITY;
    Zoom = ZOOM;
    Position = vec3(0, 0, 3);
    WorldUp = vec3(0, 1, 0);
    WorldRight = vec3(1, 0, 0);
    WorldDown = vec3(0, -1, 0);
    Center = Position + Front;
    Yaw = YAW;
    Pitch = PITCH;
    updateCameraVectors();
}

void Camera::updateCameraVectors()
{
    // calculate the new Front vector
    vec3 front;
    front.x = (float)(cos(radians(Yaw)) * cos(radians(Pitch)));
    front.y = (float)sin(radians(Pitch));
    front.z = (float)(sin(radians(Yaw)) * cos(radians(Pitch)));
    Front = normalize(front);
    // also re-calculate the Right and Up vector
    Right = normalize(cross(Front, WorldUp)); // normalize the vectors, because their length gets closer to 0 the more you look up or down which results in slower movement.
    Up = normalize(cross(Right, Front));
}
// processes input received from a mouse input system. Expects the offset value in both the x and y direction.
void Camera::ProcessMouseMovement(double xoffset, double yoffset, bool constrainPitch)
{
    xoffset *= MouseSensitivity;
    yoffset *= MouseSensitivity;
    Yaw += xoffset;
    Pitch += yoffset;
    // make sure that when pitch is out of bounds, screen doesn't get flipped
    if (constrainPitch)
    {
        if (Pitch > 89.0f)
            Pitch = 89.0f;
        if (Pitch < -89.0f)
            Pitch = -89.0f;
    }
    // update Front, Right and Up Vectors using the updated Euler angles
    updateCameraVectors();
}

void Camera::ProcessKeyboard(Camera_Movement direction, float deltaTime)
{
    float velocity = MovementSpeed * deltaTime;

    if (direction == FORWARD)
        Position = Position  + Up * velocity;
    if (direction == BACKWARD)
        Position = Position - Up * velocity;
    if (direction == LEFT)
        Position = Position - Right * velocity;
    if (direction == RIGHT)
        Position = Position + Right * velocity;
}
void Camera::Move(Camera_Movement direction, float delta) {
    if (direction == UP)
        Position = Position + Up * delta;
    if (direction == DOWN)
        Position = Position - Up * delta;
    if (direction == LEFT)
        Position = Position - Right * delta;
    if (direction == RIGHT)
        Position = Position + Right * delta;
    if (direction == FORWARD)
        Position = Position + Front * delta;
    if (direction == BACKWARD)
        Position = Position - Front * delta;
}
void Camera::camera_rotate(bool isClockWise, bool isHorizontal, float delta) {
    vec4 currentDelta(Front.x, Front.y, Front.z, 1);
    currentDelta = currentDelta * 2.0f;
    vec3 temp = Position + vec3(currentDelta);
    mat4 rotate(1);
    if (isClockWise) {
        if (isHorizontal) {
            rotate = glm::rotate(rotate, radians(delta), WorldUp);
        }
        else {
            rotate = glm::rotate(rotate, radians(delta), WorldRight);
        }
        currentDelta = rotate * currentDelta;
        
        /*
        qDebug() << "r0 " << " " << rotate[0][0] << " " << rotate[0][1] << " " << rotate[0][2] << " " << rotate[0][3];
        qDebug() << "r1 " << " " << rotate[1][0] << " " << rotate[1][1] << " " << rotate[1][2] << " " << rotate[1][3];
        qDebug() << "r2 " << " " << rotate[2][0] << " " << rotate[2][1] << " " << rotate[2][2] << " " << rotate[2][3];
        qDebug() << "r3 " << " " << rotate[3][0] << " " << rotate[3][1] << " " << rotate[3][2] << " " << rotate[3][3];
        */

        Position = temp - vec3(currentDelta);
        vec4 t2(Front.x, Front.y, Front.z, 0);
        t2 = rotate * t2;
        Front = vec3(t2.x, t2.y, t2.z);
        if ((Front - WorldUp).length() > 0.1f) {
            Right = normalize(cross(Front, WorldUp));
        }
        else {
            Right = normalize(cross(Front, WorldDown));
        }
    }
    else {
        if (isHorizontal) {
            rotate = glm::rotate(rotate, radians(-delta), WorldUp);
        }
        else {
            rotate = glm::rotate(rotate, radians(-delta), WorldRight);
        }
        currentDelta = rotate * currentDelta;
        Position = temp - vec3(currentDelta);
        vec4 t2(Front.x, Front.y, Front.z, 0);
        t2 = rotate * t2;
        Front = vec3(t2);
        if ((Front-WorldUp).length() > 0.1f) {
            Right = normalize(cross(Front, WorldUp));
        }
        else {
            Right = normalize(cross(Front, WorldDown));
        }
    }
}

mat4 Camera::GetViewMatrix()
{
    // returns the view matrix calculated using Euler Angles and the LookAt Matrix
    vec3 center = Position + Front;
    return lookAtRH(Position, center, Up);
}

vec3 Camera::ToObjectOffset()
{
    return Position;
}
void Camera::rotate_around(Camera_Movement axis, float delta) {
    vec3 center = Position + Front * 2.0f;
    
}
